### PR TITLE
External Secret Support in Composition Types

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

This PR makes the required changes to use External Secret Stores with Composition types:

- Refactors existing interfaces (e.g. `ConnectionPublisher.PublishConnection`) to use the same interface for both managed and Composition resources.
- Adds `publishConnectionDetailsTo` field to managed resource spec.
- Modifies `SecretStore` interface to support publishing/propagating for Composition types.
- Adds metadata support for Vault kv v1
- Tracks ownership of connection secrets with owner uid.
- Handles owner management for connection secrets, e.g.: 
  - cannot establish control of existing connection secret
  - existing secret is not owned by UID ...
  - refusing to modify unowned secret of type ...

Partially Fixes https://github.com/crossplane/crossplane/issues/2881

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

See https://github.com/crossplane/crossplane/pull/2940

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/common/v1/resource.go
Comment (negz) [inline apis/common/v1/resource.go:145]: Cross reference to https://github.com/crossplane/crossplane/issues/2929 - might be worth considering before we introduce this new field.

Comment (turkenh) [reply apis/common/v1/resource.go:145]: Responded there: https://github.com/crossplane/crossplane/issues/2929#issuecomment-1060493213

Comment (negz) [reply apis/common/v1/resource.go:145]: Either way, it's probably worth documenting here and/or on WriteConnectionSecretToRef what happens when both are set / what their relationship is.

File: pkg/connection/fake/mocks.go
Comment (negz) [inline pkg/connection/fake/mocks.go:37]:
```suggestion
	DeleteConnectionDetailsFn func(ctx context.Context, i store.Secret, conn managed.ConnectionDetails) error
```

Is it worth changing the names of these methods to match the new args/return values?

File: pkg/connection/manager.go
Comment (negz) [inline pkg/connection/manager.go:188]: I'm hazy on how this used to work, but I want to make sure that we won't face a security regression if we don't implement this. I think we might have been partly using `ConnectionSecretMustBeControllableBy` to try to avoid malicious folks abusing Crossplane by asking it to overwrite a secret that it shouldn't - for example to propagate `EvilXR`'s connection details to `VictimXRC`'s connection secret.

File: pkg/connection/manager_test.go
Comment (negz) [inline pkg/connection/manager_test.go:25]: Nit - extra whitespace between these imports.

File: pkg/reconciler/managed/api.go
Comment (negz) [inline pkg/reconciler/managed/api.go:103]: Nice, I like making these take a narrower interface.

File: pkg/resource/interfaces.go
Comment (negz) [inline pkg/resource/interfaces.go:246]: I don't think that would be needed under the new design. I'm pretty sure the difference between `Connection` and `LocalConnection` was that one had a namespaced ref (e.g. with name and namespace) while the other was just a name (implying writing to the claim's namespace).

**Review by negz — COMMENTED** [review]
Not approving just yet because this needs a rebase now that #322 is merged, and it seems like there's a few TODOs left to address but overall this looks good.

File: pkg/connection/fake/mocks.go
Comment (turkenh) [reply pkg/connection/fake/mocks.go:37]: Hmm, checking this again, I didn't want to leak `managed.ConnectionDetails` in `connection.store` package because this is the interface that SecretStore implementations need to satisfy and these are just KeyValues for them, i.e. connection details does not mean something for a secret store.
Refactored accordingly.

File: pkg/resource/interfaces.go
Comment (turkenh) [reply pkg/resource/interfaces.go:246]: Yeah, I can confirm this with my testing so far.

**Comment by turkenh** [comment]
@negz this is ready for another pass.

I've resolved previous todo's and had to make some further changes especially to handle ownership and related checks. Tested end to end with https://github.com/crossplane/crossplane/pull/2940 and that PR has the latest changes.

File: apis/common/v1/connection_details.go
Comment (negz) [inline apis/common/v1/connection_details.go:69]: Presuming this is used to track ownership, should we add a note on the `ConnectionSecretMetadata` type that Secret Store implementations must implement label support in order to track secret ownership?

File: apis/common/v1/connection_details.go
Comment (negz) [inline apis/common/v1/connection_details.go:26]: This seems like it would duplicate Kubernetes Secret controller references in some cases. I suppose we need to go this path if we want to be consistent though, given that some Kubernetes Secrets might live in different clusters, and obviously controller references don't exist in Vault. Perhaps a comment noting that would help future readers understand why we use labels for this?

File: pkg/connection/manager.go
Comment (negz) [inline pkg/connection/manager.go:145]: Would using `SecretMustBeOwnedBy` make sense here too, to make sure we don't delete a secret we don't own?

File: pkg/connection/manager_test.go
Comment (negz) [inline pkg/connection/manager_test.go:214]:
```suggestion
		"CannotPublishTo": {
```

Nit

File: pkg/reconciler/managed/api.go
Comment (negz) [inline pkg/reconciler/managed/api.go:120]: Huh, weird, I guess the comment for this function was previously inaccurate in that it said it would be a no-op if the secret already existed.

File: pkg/reconciler/managed/publisher_test.go
Comment (negz) [inline pkg/reconciler/managed/publisher_test.go:25]: Nit: I think we usually import this as `xpv1`.

**Review by negz — APPROVED** [review]
This looks good to me - thank you! I've added a few comments but none of them are blocking. I won't merge right now just in case you want to follow up on any of the comment, but feel free to have @muvaf or someone else with access merge it if you don't want to wait for me to come online.

File: pkg/connection/manager.go
Comment (turkenh) [reply pkg/connection/manager.go]: Yes, good idea, done!

File: pkg/connection/manager.go
Comment (turkenh) [inline pkg/connection/manager.go]: Removed taking ownership of an unowned resource assuming this was for some sort of backward compatibility thing. This simplified the logic here a lot.

**Comment by turkenh** [comment]
> This looks good to me - thank you! I've added a few comments but none of them are blocking. I won't merge right now just in case you want to follow up on any of the comment, but feel free to have @muvaf or someone else with access merge it if you don't want to wait for me to come online.

Thanks a lot @negz for the review and approval. Resolved your last set of comments. 

@muvaf would you mind hitting the merge button? it would also be great if you could review the last two commits after @negz 's approval.

**Comment by muvaf** [comment]
@turkenh I don't have all the context so just hit the merge button since it's already got the approval. Thanks!

**Pull request merged**
